lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1595,6 +1649,169 @@ class BuildBBRegs : public FuncOpPartialLoweringPattern {
+        auto reg = createReg(getComponentState(), rewriter, value.getLoc(),
+                             name, width);
+        getComponentState().addPipelineReg(stage, reg, i);
+
+        // Note that we do not RAUW here as in BuildBBRegs. Instead, we wait
+        // until after BuildOpGroups, and RAUW inside BuildPipelineGroups, once
+        // the pipeline register created here has been assigned.
+      }
+    });
+    return success();
+  }
+};
+
+/// Builds groups for assigning registers for pipeline stages.
+class BuildPipelineGroups : public FuncOpPartialLoweringPattern {
+  using FuncOpPartialLoweringPattern::FuncOpPartialLoweringPattern;
+
+  LogicalResult
+  PartiallyLowerFuncToComp(mlir::FuncOp funcOp,
+                           PatternRewriter &rewriter) const override {
+    for (auto pipeline : funcOp.getOps<staticlogic::PipelineWhileOp>())
+      for (auto stage :
+           pipeline.getStagesBlock().getOps<staticlogic::PipelineStageOp>())
+        if (failed(buildStageGroups(stage, rewriter)))
+          return failure();
+
+    return success();
+  }
+
+  LogicalResult buildStageGroups(staticlogic::PipelineStageOp stage,
+                                 PatternRewriter &rewriter) const {
+    // Collect pipeline registers for stage.
+    auto pipelineRegisters = getComponentState().getPipelineRegs(stage);
+
+    // Collect group names for the prologue or epilogue.
+    SmallVector<StringAttr> prologueGroups;
+    SmallVector<StringAttr> epilogueGroups;
+    auto whileOp = stage->getParentOfType<staticlogic::PipelineWhileOp>();
+
+    // For each registered stage result value.
+    auto stageResults = stage.getBodyBlock().getTerminator()->getOperands();
+    for (size_t i = 0, e = stageResults.size(); i < e; ++i) {
+      auto value = stageResults[i];
+
+      // Get the pipeline register for that result.
+      auto pipelineRegister = pipelineRegisters[i];
+
+      // Get the evaluating group for that value.
+      auto evaluatingGroup = getComponentState().getEvaluatingGroup(value);
+
+      // Remember the final group for this stage result.
+      calyx::GroupOp group;
+
+      // Stitch the register in, depending on whether the group was
+      // combinational or sequential.
+      if (auto combGroup =
+              dyn_cast<calyx::CombGroupOp>(evaluatingGroup.getOperation())) {
+        // Create a sequential group and replace the comb group.
+        PatternRewriter::InsertionGuard g(rewriter);
+        rewriter.setInsertionPoint(combGroup);
+        group = rewriter.create<calyx::GroupOp>(combGroup.getLoc(),
+                                                combGroup.getName());
+        rewriter.cloneRegionBefore(combGroup.getBodyRegion(), group.getBody());
+        group.getBodyRegion().back().erase();
+        rewriter.eraseOp(combGroup);
+
+        // Stitch evaluating group to register.
+        buildAssignmentsForRegisterWrite(getComponentState(), rewriter, group,
```

Comment:
Alternatively to the above (or alongside); we could in builder functions like this automatically rewrite a combinational group to a sequential group in cases where we're building sequential logic into them.

Reply:
I see what you're getting at, but having that kind of side-effect in a method like this might be suprising to some. I'm thinking I will pull out helpers as you suggested above, and keep those orthogonal to buildAssignmentsForRegisterWrite.